File: 437/437.v1.cpp
#include <cstdlib>

struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
 	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 };

class Solution {
private:
	int pathSumContinuous(TreeNode* root, int sum) {
		if (!root)
			return 0;
		
		return
			(root->val == sum) +                             // Path ending on current node
			pathSumContinuous(root->left, sum - root->val) + // Paths starting from a previous
			                                                 // node and ending on any node in
			                                                 // left child tree
			pathSumContinuous(root->right, sum - root->val); // Paths starting from a previous
			                                                 // node and ending on any node in
			                                                 // right child tree
	}
public:
	int pathSum(TreeNode* root, int sum) {
		if (!root)
			return 0;
		return
			pathSumContinuous(root, sum) + // Paths starting from current node
			pathSum(root->left, sum) + // Paths starting from any node in left child tree
			pathSum(root->right, sum); // Paths starting from any node in right child tree
	}
};

File: 400-499/437/437.v1.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func pathSumContinuous(root *TreeNode, sum int) int {
	if root == nil {
		return 0
	}

	path := 0

	// Path ending on current node
	if root.Val == sum {
		path += 1
	}

	// Paths starting from a previous node and ending on any node in left child tree
	path += pathSumContinuous(root.Left, sum-root.Val)

	// Paths starting from a previous node and ending on any node in right child tree
	path += pathSumContinuous(root.Right, sum-root.Val)

	return path
}

func pathSum(root *TreeNode, sum int) int {
	if root == nil {
		return 0
	}

	path := 0

	// Paths starting from current node
	path += pathSumContinuous(root, sum)

	// Paths starting from any node in left child tree
	path += pathSum(root.Left, sum)

	// Paths starting from any node in right child tree
	path += pathSum(root.Right, sum)

	return path
}
